rust/src/io/writer.rs

```diff
@@ -0,0 +1,225 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow_array::cast::as_struct_array;
+use arrow_array::{Array, ArrayRef, RecordBatch, StructArray};
+use arrow_schema::DataType;
+use async_recursion::async_recursion;
+use tokio::io::AsyncWriteExt;
+
+use crate::arrow::*;
+use crate::datatypes::{is_fixed_stride, Field, Schema};
+use crate::encodings::binary::BinaryEncoder;
+use crate::encodings::Encoder;
+use crate::encodings::{plain::PlainEncoder, Encoding};
+use crate::format::{Manifest, Metadata, PageInfo, PageTable, MAGIC, MAJOR_VERSION, MINOR_VERSION};
+use crate::io::object_writer::ObjectWriter;
+use crate::{Error, Result};
+
+/// FileWriter writes Arrow Table to a file.
+pub struct FileWriter<'a> {
+    object_writer: ObjectWriter,
+    schema: &'a Schema,
+    batch_id: i32,
+    page_table: PageTable,
+    metadata: Metadata,
+}
+
+impl<'a> FileWriter<'a> {
+    pub fn new(object_writer: ObjectWriter, schema: &'a Schema) -> Self {
+        Self {
+            object_writer,
+            schema,
+            batch_id: 0,
+            page_table: PageTable::default(),
+            metadata: Metadata::default(),
+        }
+    }
+
+    /// Write a [RecordBatch] to the open file.
+    ///
+    /// Returns [Err] if the schema does not match with the batch.
+    pub async fn write(&mut self, batch: &RecordBatch) -> Result<()> {
+        for field in &self.schema.fields {
+            let column_id = batch.schema().index_of(&field.name)?;
+            let array = batch.column(column_id);
```

Comment:
should we just implement `RecordBatch.column(&str)` ?

Reply:
You mean extend it? sure, why not :) 

maybe we should have a `lance::preclude::*`

Reply:
lemme do it in the follow up PRs?